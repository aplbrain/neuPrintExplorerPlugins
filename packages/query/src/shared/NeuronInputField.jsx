import React from 'react';
import PropTypes from 'prop-types';
import AsyncSelect from 'react-select/async';
import InputLabel from '@material-ui/core/InputLabel';
import { withStyles } from '@material-ui/core/styles';

const styles = theme => ({
  regexWarning: {
    fontSize: '0.9em'
  },
  select: {
    fontFamily: theme.typography.fontFamily,
    margin: '0.5em 0 1em 0'
  }
});

class NeuronInputField extends React.Component {
  handleChange = selected => {
    const { onChange } = this.props;
    if (selected && selected.value) {
      onChange(selected.value);
    } else {
      onChange(selected);
    }
  };

  handleKeyDown = event => {
    const { handleSubmit } = this.props;
    // submit request if user presses enter
    if (event.keyCode === 13) {
      event.preventDefault();
      handleSubmit();
    }
  };

  fetchOptions = inputValue => {
    const { dataSet } = this.props;

    const convertedInput = parseInt(inputValue, 10);

    let bodyId = -1;

    if (!Number.isNaN(convertedInput)) {
      bodyId = convertedInput;
    }

    // query neo4j
    const cypherString = `MATCH (neuron :Neuron)
    WHERE neuron.bodyId = ${bodyId}
    OR toLower(neuron.type) CONTAINS toLower('${inputValue}')
    OR toLower(neuron.instance) CONTAINS toLower('${inputValue}')
    RETURN neuron.bodyId AS bodyid, neuron.type AS type, neuron.instance AS instance
    ORDER BY neuron.instance`;

    const body = JSON.stringify({
      cypher: cypherString,
      dataSet
    });

    const settings = {
      headers: {
        'content-type': 'application/json',
        Accept: 'application/json'
      },
      body,
      method: 'POST',
      credentials: 'include'
    };

    const queryUrl = '/api/custom/custom';

    return fetch(queryUrl, settings)
      .then(result => result.json())
      .then(resp => {
        // sort the results in the data key. Need to split out instances, types
        // and bodyids into separate categories then load them in different
        // sections of the drop down.

        const bodyIds = new Set();
        const types = new Set();
        const instances = new Set();

        resp.data.forEach(item => {
          if (item[0]) {
            bodyIds.add(item[0].toString());
          }
          if (item[1]) {
            types.add(item[1]);
          }
          if (item[2]) {
            instances.add(item[2]);
          }
        });

        const options = [];

        if (types.size) {
          options.push({
            label: 'Types',
            options: [...types].sort().slice(0,9).map(item => ({ value: item, label: item }))
          });
        }
        if (instances.size) {
          options.push({
            label: 'Instances',
            options: [...instances].sort().slice(0,9).map(item => ({ value: item, label: item }))
          });
        }
        if (bodyIds.size) {
          options.push({
            label: 'Body Ids',
            options: [...bodyIds].sort((a,b) => a - b).slice(0,9).map(item => ({ value: item, label: item }))
          });
        }

        return options;
      });
  };

  render() {
    const { value, classes } = this.props;
    const selectValue = value ? { label: value, value } : null;
    return (
      <React.Fragment>
        <InputLabel htmlFor="select-multiple-chip">
          Neuron Instance, Type or BodyId (optional)
        </InputLabel>
        <AsyncSelect
          className={classes.select}
          placeholder="Type or Paste text for options"
          value={selectValue}
          isClearable
          loadOptions={this.fetchOptions}
          onChange={this.handleChange}
        />
      </React.Fragment>
    );
  }
}

NeuronInputField.propTypes = {
  classes: PropTypes.object.isRequired,
  handleSubmit: PropTypes.func.isRequired,
  onChange: PropTypes.func,
  dataSet: PropTypes.string.isRequired,
  value: PropTypes.string
};

NeuronInputField.defaultProps = {
  onChange: () => {},
  value: ''
};

export default withStyles(styles)(NeuronInputField);
