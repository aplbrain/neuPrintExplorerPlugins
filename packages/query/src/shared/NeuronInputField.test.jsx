import NeuronInputField from './NeuronInputField';

let wrapper;
let textField;

const { React, enzyme, renderer } = global;

const processRequest = jest.fn();
const addNeuronInstance = jest.fn();

const component = (
  <NeuronInputField
    onChange={addNeuronInstance}
    value='test'
    handleSubmit={processRequest}
  />
);

describe('NeuronInputField', () => {
  beforeAll(() => {
    wrapper = enzyme.mount(component);
    textField = wrapper.find('TextField').at(0);
  });

  it('renders correctly', () => {
    const pluginView = renderer.create(component).toJSON();
    expect(pluginView).toMatchSnapshot();
  });

  describe('when enter key is pressed', () => {
    it('should call the callback', () => {
      const preventDefault = jest.fn();
      textField.props().onKeyDown({ keyCode: 13, preventDefault });
      expect(preventDefault).toHaveBeenCalledTimes(1);
      expect(processRequest).toHaveBeenCalledTimes(1);
    });
  });

});

