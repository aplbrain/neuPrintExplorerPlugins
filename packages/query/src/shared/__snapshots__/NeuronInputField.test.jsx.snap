// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NeuronInputField renders correctly 1`] = `
<div
  className="MuiFormControl-root-2 MuiFormControl-fullWidth-5"
  onKeyDown={[Function]}
>
  <label
    className="MuiFormLabel-root-17 MuiFormLabel-filled-21 MuiInputLabel-root-6 MuiInputLabel-formControl-11 MuiInputLabel-animated-14 MuiInputLabel-shrink-13"
    data-shrink={true}
  >
    Neuron Instance, Type or BodyID (optional)
  </label>
  <div
    className="MuiInputBase-root-37 MuiInput-root-24 MuiInput-underline-28 MuiInputBase-fullWidth-46 MuiInput-fullWidth-31 MuiInputBase-formControl-38 MuiInput-formControl-25 MuiInputBase-multiline-45 MuiInput-multiline-30"
    onClick={[Function]}
  >
    <div
      className="MuiPrivateTextarea-root-54"
    >
      <textarea
        aria-hidden="true"
        className="MuiPrivateTextarea-textarea-55 MuiPrivateTextarea-shadow-56"
        readOnly={true}
        rows="1"
        tabIndex={-1}
        value=""
      />
      <textarea
        aria-hidden="true"
        className="MuiPrivateTextarea-textarea-55 MuiPrivateTextarea-shadow-56"
        readOnly={true}
        rows={1}
        tabIndex={-1}
        value="test"
      />
      <textarea
        aria-invalid={false}
        className="MuiPrivateTextarea-textarea-55 MuiInputBase-input-47 MuiInput-input-32 MuiInputBase-inputMultiline-49 MuiInput-inputMultiline-34"
        disabled={false}
        onBlur={[Function]}
        onChange={[Function]}
        onFocus={[Function]}
        required={false}
        rows={1}
        style={
          Object {
            "height": 19,
          }
        }
        value="test"
      />
    </div>
  </div>
</div>
`;
